pace-util/pace/util/checkpointer/thresholds.py

```diff
@@ -0,0 +1,155 @@
+    pass
+
+
+@dataclasses.dataclass
+class Threshold:
+    relative: float
+    absolute: float
+
+    def merge(self, other: "Threshold") -> "Threshold":
+        """
+        Provide a threshold which is always satisfied
+        if both input thresholds are satisfied.
+
+        This is generally a less strict threshold than either input.
+        """
+        return Threshold(
+            relative=max(self.relative, other.relative),
+            absolute=max(self.absolute, other.absolute),
+        )
+
+
+@dataclasses.dataclass
+class SavepointThresholds:
+    savepoints: Dict[SavepointName, List[Dict[VariableName, Threshold]]]
+
+
+def cast_to_ndarray(array: ArrayLike) -> np.ndarray:
+    if isinstance(array, Quantity):
+        array = array.data
+    if isinstance(array.data, np.ndarray):
+        return array.data
+    else:
+        return array
+
+
+class ThresholdCalibrationCheckpointer(Checkpointer):
+    """
+    Calibrates thresholds to be used by a ValidationCheckpointer.
+
+    Does this by recording the minimum and maximum values seen across trials,
+    and using them to derive the maximum relative and absolute error one could
+    have across any pair of trials, then multiplying this by a user-provided factor.
+    """
+
+    def __init__(self, factor: float = 1.0):
+        """
+        Args:
+            factor: set thresholds equal to this factor of the maximum error
+                seen across trials
+        """
+        self._minimums: Mapping[
+            SavepointName, List[Mapping[VariableName, np.ndarray]]
+        ] = collections.defaultdict(list)
+        self._maximums: Mapping[
+            SavepointName, List[Mapping[VariableName, np.ndarray]]
+        ] = collections.defaultdict(list)
+        self._factor = factor
+        self._abs_sums: Mapping[
+            SavepointName, List[Mapping[VariableName, np.ndarray]]
+        ] = collections.defaultdict(list)
+        self._n_trials = 0
+        self._n_calls: Mapping[SavepointName, int] = collections.defaultdict(int)
+
+    def __call__(self, savepoint_name, **kwargs):
+        """
+        Record values for a savepoint.
+
+        Args:
+            savepoint_name: name of the savepoint
+            **kwargs: data for the savepoint
+        """
+        i_call = self._n_calls[savepoint_name]
+        if len(self._minimums[savepoint_name]) < i_call + 1:
+            self._minimums[savepoint_name].append(
+                collections.defaultdict(lambda: np.inf)
+            )
+            self._maximums[savepoint_name].append(
+                collections.defaultdict(lambda: -np.inf)
+            )
+            self._abs_sums[savepoint_name].append(collections.defaultdict(lambda: 0.0))
+        for varname, array in kwargs.items():
+            array: np.ndarray = cast_to_ndarray(array)
+            self._minimums[savepoint_name][i_call][varname] = np.minimum(
+                self._minimums[savepoint_name][i_call][varname], array
+            )
```

Comment:
I know this probably just worked, but I don't understand how you can have `[i_call][varname]`. I thought that line 93 gives you one extra dict, but the second one?

Reply:
the `i_call` part is for list indexing, not dictionary lookup. It's a dictionary (over savepoint) of list (over call count) of dictionary (over variable name).

I just added a line comment to this effect in the init.